### DISCUSSION TITLE
Comparision with django-cotton

### DISCUSSION THREAD

**Discussion by EmilStenstrom** [django-components/django-components#552] [Show and tell]
https://github.com/wrabit/django-cotton?tab=readme-ov-file#comparison-with-other-packages

**Comment by JuroOravec** [comment]
The ability to define Python objects like lists within the template is really nice! ([See Passing Python data types](https://github.com/wrabit/django-cotton#passing-python-data-types))

**Comment by JuroOravec** [comment]
Btw, curiosity got the better of me, and I got a working algo for safely evaluating expressions, so theoretically we could have a similar feature to what django-cotton does python expressions within HTML.

E.g. in Cotton, they can pass a static list to a component like so:
```html
<c-select name="q1" :options="['yes', 'no', 'maybe']" />
```

Django-cotton uses [`ast.literal_eval`](https://stackoverflow.com/questions/15197673/using-pythons-eval-vs-ast-literal-eval), which means that the expressions cannot contain variables. I took slightly different approach, by first searching the AST for variables and checking for permitted syntax.

The result is that our expressions can contain variables. (For context, unlike plain `eval/exec`, this forbids assignments or any other statements. And I also disabled function calls, to stick with the theme of Django templates).

So, we could have something like this (exaggarated example):

```django
{% component "my_comp"
	:items="[a for a in [1, 2, *(list_a if 1 > 2 else list_b) ] ]"
	:bool_val="not is_active"
	:attrs="{**dict_1, **dict_2}"
%}
{% endcomponent %}
```

This would fit nicely with what I talked about in https://github.com/EmilStenstrom/django-components/issues/539.

**Reply by EmilStenstrom** [reply]
Exciting! I think this feature is very nice :)

So colon in front of a variable means "this has executable contents"? Maybe. Here's some alternatives to consider:
* {% component items="{{ a for a in [1, 2, 3, 4] if a % 2 == 0 }}" %}
* {% component items=eval("a for a in [1, 2, 3, 4] if a % 2 == 0)" %}

I expect this to be really useful in practice!

**Reply by JuroOravec** [reply]
Great! Of the 3, I'm on the fence between the colon prefix and double-curly braces.

At first, my preference was for something that marks the dynamic expression that's outside of the expression itself. The double-curly braces could also work, tho at first it feels could be harder to read with dicts:

`{% component items="{{ {'id': a, "abc": "def"} }}" %}`

On the other hand, double curly braces are well known in Django templates, whereas the colon prefix would be foreign. And the more I think about it, the more I'm in favour of the double curly braces.

---

I'm not in favour of something like `eval()`, because, after some thought, I think the expressions should allow also function calls. So something that looks like a function could be confusing/

Specifically I want to allow function calls to make life easier in the case of when one wants to process data obtained from a slot:

```django
{# We cannot modify List and Button components, they are 3rd party #}
{% component "list" %}
	{% fill data="item" %}
		{# I want to do some complex processing on `item` before I pass it to the button #}
		{% component "button"
			attrs="{{ process_item_to_attrs(item) }}"
		%}
		{% endomponent %}
	{% endfill %}
{% endcomponent %}
```

Otherwise, in the example above, one would be forced to create a subcomponent just to wrap the button component and process the data from the slot.

**Reply by EmilStenstrom** [reply]
Curly brackets it is! I think it's the one that makes the most sense to beginners as well. 

Connected to another thread, maybe we could preserve the returned types too? That means...
```django
{% component "button" items="{{ ['yes', 'no'] }}" %}
``` 
...would be a way to pass an actual python list from the template to the component?

**Reply by JuroOravec** [reply]
Yup, that's how I've been thinking about too! Played around it today, got it working also with the tests. So I'll make an MR once the TagFormatter is out of the way.

Since we support both positional and keyword args, I made it work for both:
```django
{% component 'test'
    "{{ [a for a in [1, 2, *([3, 4] if 1 > 2 else x)]] }}"    <-- positional
    "{{ 'hallo ' + name + ' gut' }}"                          <-- positional
    bool_var="{{ not is_active or (False or 0) }}"            <-- keyword
    dict_var="{{ {**dict1, 'c': 'd', **dict2} }}"             <-- keyword
/ %}
```

Moreover, I made these expressions into a stanalone tag, I called it `expr`. So that this syntax can be used also elsewhere.
```django
{% expr "{{ 'hallo ' + name + ' gut' }}" %}
```

Note on security:

For one, I enabled function calls, as mentioned above. Moreover, to be able to do basic python data transformations like. `str(2)`,  I had to expose Python built-ins to the expression code. I think that's fine, after all the code React templates also has full access.

But since Django runs on the server, it also means that these expressions are definitely not safe to run with user-generated content.  User could send a template with `{% expr "{{ exit() }} " %}` and shut the server down.

So, for one, I'm writing this as a reminder for myself, to add a warning to documentation, when I come around to it. And I think the combo of function calls + access to globals should be opt-in, and OFF by default.

**Reply by EmilStenstrom** [reply]
Hmm... that does indeed sound dangerous. Following Django's lead here would mean you can't do arbitrary python, only things that would be valid in a django template. Personally I think that would make a lot more sense. Django has this philosophy that templates should contain as little logic as possible.

Yes, I realize that this means my example above wouldn't work, because what I wrote isn't valid in a Django template. But I think making things work like they would think it works is the best design choice almost always...

So I would suggest we strip it back to only work like in a template...

**Reply by JuroOravec** [reply]
Do I understand correctly, that it would then work as a regular `{{ }}` in a Django template? that you could apply filters like in a normal template, e.g:
```django
{% component 'test'
    "{{ my_list|join:',' }}"                 <-- positional
    title="{{ my_name | title }}"             <-- keyword
/ %}
```

If so, then that's fine with me, as I can still release the expressions logic as filter/tag in a standalone package, so I can use it with the UI templates, and have something like this:

```django
{% component 'test'
    "{{ '[a for a in [1, 2, *([3, 4] if 1 > 2 else x)]]' | expr }}"    <-- positional
    bool_var="{{ 'not is_active or (False or 0)' | expr }}"            <-- keyword
/ %}
```

Although the above won't be very practical with text, since both single and double quotes are already used up, so all strings would have to have their quotes escaped. But that's my problem, as I want to use those expressions 😄 And most of the time the expressions will be used to work with bools/lists/dicts anyway, so shouldn't be much of an issue.

But I agree that for django_components, it makes more sense to be aligned with Django. 

Anyway, I should be done with the JS part of the UI library in about a week, and then I'm moving onto Django components. And while this feature doesn't need to be implemented by then (after all, I have all the proposed features on my fork), I'd like the API to be agreed on, so I can write the components with the already new syntax!

**Reply by EmilStenstrom** [reply]
Yes, I'm thinking that everything within double brackets would behave as if it were outside of the component tags. That is, with |join:"," and dot notation to access stuff. Is there a simple way to not have to reimplement djangos template language, and just execute a string as if it were in a normal template, with the same context as the component tag?

An expr tag sounds like a fun thing, but I agree that it should be outside of django_components!

**Reply by JuroOravec** [reply]
Had a quick look just now, should be possible:
1. [This](https://github.com/django/django/blob/a57596e443ecb67140e1a9fc0f0e87446b2d0174/django/template/base.py#L394) is the logic that gets executed when Django template comes across `{{ }}` tags 
2. The contents of the `{{ }}` tag are then passed to `VariableNode` [here](https://github.com/django/django/blob/a57596e443ecb67140e1a9fc0f0e87446b2d0174/django/template/base.py#L490)
3. `VariableNode` then uses `FilterExpression.resolve` to evaluate the contents with a given `Context` [here](https://github.com/django/django/blob/a57596e443ecb67140e1a9fc0f0e87446b2d0174/django/template/base.py#L1067)
4. The resolution logic is then ultimately [here](https://github.com/django/django/blob/a57596e443ecb67140e1a9fc0f0e87446b2d0174/django/template/base.py#L715).

So I expect that it should be enough to create `FilterExpression` out of the tag contents, and then call `FilterExpression.resolve()` with the context to get the resolved values, which hopefully won't be stringified. And we can also call `Parser.parse()` on the contents to have the same syntax validation as normal Django templates (e.g. would raise an error if the tag is empty, [like here](https://github.com/django/django/blob/a57596e443ecb67140e1a9fc0f0e87446b2d0174/django/template/base.py#L484)).

**Reply by EmilStenstrom** [reply]
@JuroOravec that sounds awesome! :D

**Reply by JuroOravec** [reply]
So I migrated the expression logic into it's own package, [django_expr](https://github.com/JuroOravec/alpinui/tree/main/packages/django-expr).

While doing so, I realized one thing, and that's that filters don't have access to the `Context` object. Only to the value preceding the filter, and optionally the argument after the colon `:`.

This would make dynamic expressions much less useful as it wouldn't be possible to use them to e.g. join two (or more) dicts (or other types).

What do you think of allowing both filters `{{ }}` and tags `{% %}`? So one could do both this:

```django
{% component 'test'
    title="{{ my_name | title }}"
/ %}
```

and this

```django
{% component 'test'
    title="{% expr 'person.first_name + \' \'+ person.last_name' %}"
/ %}
```

**Reply by JuroOravec** [reply]
Ah, nevermind, I tried to implement the nested `{% %}` tags, but the way Django is written, I would have to make changes to the `Template` class. So users would have to use a custom `Template` class to be able to achieve the above, and I don't want to go down that route.

So closest to React/Vue I can have with Django templates is placing the template tags on separate lines:

```django
{% expr 'person.first_name + \' \'+ person.last_name' as title %}

{% component 'test' title=title / %}
```

**Reply by EmilStenstrom** [reply]
Having expr separate is probably the right call. In a typical Django project, you would move that expression to the view, and have a ViewModel (the context dict) that has all the values you need in the template pre-populated.

Another API to consider is if we would allow `{% component 'test' title="{{ person.first_name}} {{ person.last_name}}" / %}`? Maybe overkill as well?

**Reply by JuroOravec** [reply]
> Having expr separate is probably the right call. In a typical Django project, you would move that expression to the view, and have a ViewModel (the context dict) that has all the values you need in the template pre-populated.

Agree. I mean, it's a minor thing, but I initially wanted expressions inside templates, because, having experience from Vue, it allows to give a bit more meaning to the code. As in, if I expose a variable to the template, that's noteworthy. And when I need to negate that variable or slightly modify it, I know it's trivial, because that expression doesn't even get it's own variable. On the other hand, if I expose `variable` and `notVariable` and `filteredVariable`, it's harder to see what's really important.

But, after all, I can still achieve that distinction with the example in my previous message.

> Another API to consider is if we would allow `{% component 'test' title="{{ person.first_name}} {{ person.last_name}}" / %}`? Maybe overkill as well?

I considered it too, but then IMO it would make sense to strictly render the inner template as string. On the other hand, if we allow only a single pair of `{{ }}`, then we can evaluate it without forcing the result to be stringified.

**Reply by EmilStenstrom** [reply]
Again in the Django world, they would argue that allowing expressions in templates means that people will misuse it and add business logic there. This is not a problem with a bunch of experienced developers, but  with more junior developers it could lead to very messy code. 

All in all, I think having expr as an external lib, and using the "as variable" convention, is a great way to navigate around this.

Avoiding stringification is a great point.

**Reply by oliverhaas** [reply]
@JuroOravec I really like your approach for the safe evaluation of the expression. I have been wondering for a while how one could do this, and using the ast looks pretty solid to me. Didn't actually occur to me....

**Reply by JuroOravec** [reply]
@oliverhaas If you're interested in that, have a look at https://github.com/JuroOravec/alpinui/blob/main/packages/django-expr/README.md :) By default it doesn't allow function calls, but that can be configured via [`EXPR_ALLOW_FUNC_CALLS`](https://github.com/JuroOravec/alpinui/blob/main/packages/django-expr/src/django_expr/expression.py#L209)

**Reply by aehlke** [reply]
Is the django-cotton style possible yet? Thanks

**Reply by JuroOravec** [reply]
@aehlke Not yet, but with the addition of [extensions](https://django-components.github.io/django-components/latest/concepts/advanced/extensions/), we're getting closer. But there's still a few blockers:
- Fixing the support for literal lists and dictionaries within templates and migrating the parser to a Rust script.
  - I was working on this before getting side-tracked with the performance-related issues. I reckon I'll get back to it in 1-2 weeks.
- Clean up how django-components handles and loads Templates.
  - Right now it feels like a bit of a mess. This is where we'd add an [extension hook](https://django-components.github.io/django-components/latest/reference/extension_hooks/) to pre-process templates, and this is what will make the django-cotton style possible.
  - It's all a bit more complex and intertwined:
    1. When a component's template is defined as a file path or as inlined string, those cases are simple - we just load the content as plain string, and we can pre-process that.
    2. Components also support `get_template_name()` and `get_template()` (which IMO I would deprecate and remove in v1). These two make the implementation harder, because it means that each time a component is rendered, it may have a different Template.
    3. And so, when the template is defined as a file path or as an inlined string, we can pre-process the template only once, on the first use of the component
    4. But when using `get_template_name()` and `get_template()`, the same pre-processing will have to be called on each time the component is rendered. Alternatively, we could say that the pre-processing would NOT be applied when using these methods, to encourage people to use `Component.template` and `Component.template_file` instead.
    5. So that's why I want to first clean up the logic surrounding the templates first.

After these, it will be possible to write an extension that would pre-process the templates.

And as always, if you want to see some features sooner, be sure to support this project!

**Reply by EmilStenstrom** [reply]
@aehlke Thanks for letting us know that this is something you would want. Is this a reason to move from django-cotton to django-components? Do you have a project in mind for migration?

**Reply by aehlke** [reply]
@EmilStenstrom just evaluating current django landscape as I prepare to resume working in it after several long years. not a current user of this or cotton. thanks

**Comment by oliverhaas** [comment]
For me personally, the syntax of django-cotton itself is a nice feature, which leads me to the following question:

@EmilStenstrom Would django-components be open to include an optional cotton-like html tag syntax? We've already started implemented something along those lines in one of our projects to try it out. It's mostly a matter of taste, of course. Upsides of custom html tags are mainly formatting and intellisense regarding css classes by default. I'm using vscode, and I'm also checking out whether (automatically generated!) code snippets and html data providers would help the DX.

**Reply by JuroOravec** [reply]
@oliverhaas I like the idea, simply because from Vue I'm used to categorizing what I see in the template into 3 buckets:
1. HTML element if it's wrapped in `<>` and kebab-case, e.g. `<div>`
2. Custom element if it's wrapped in `<>` and PascalCase, e.g. `<MyButton />`
3. Something else being inserted if I see `{{ }}` (or `{% %}` for Django)

However, could you describe more what you have in mind? E.g. one way could be to simply replace all `{% %}` with `<>` in case of components, like below. Or would it entail anything else?

```django
{% component "my_comp" abc=123 xx="{% expr '1 + 1' %}" %}
{% endcomponent %}

{% component "my_comp" abc=123 xx="{% expr '1 + 1' %}" / %}

{# shorthand syntax #}
{% my_comp abc=123 xx="{% expr '1 + 1' %}" %}
{% endmy_comp %}

{% my_comp abc=123 xx="{% expr '1 + 1' %}" / %}
```

Could be written as:
```django
<component "my_comp" abc=123 xx="{% expr '1 + 1' %}">
</component>

<component "my_comp" abc=123 xx="{% expr '1 + 1' %}" />

{# shorthand syntax - NOTE: underscore replaced with dash #}
<my-comp abc=123 xx="{% expr '1 + 1' %}">
</my-comp>

<my-comp abc=123 xx="{% expr '1 + 1' %}" />
```

Or maybe PascalCase'd

```django
<Component "my_comp" abc=123 xx="{% expr '1 + 1' %}">
</Component>

<Component "my_comp" abc=123 xx="{% expr '1 + 1' %}" />

{# shorthand syntax #}
<MyComp abc=123 xx="{% expr '1 + 1' %}">
</MyComp>

<MyComp abc=123 xx="{% expr '1 + 1' %}" />
```

**Reply by EmilStenstrom** [reply]
Great reply Juro! I think that switching from curly brackets to some other tags should be something you do in the template engine, not in a library like django_components. 

One goal of django_components is that projects that use it should feel natural to Django-developers from the first time they see the code. I think `{% component "my_comp" abc=123 xx="{% expr '1 + 1' %}" / %}" / %}` just makes sense, but I think `<MyComp abc=123 xx="{% expr '1 + 1' %}" />` would generally confuse people, because HTML tags should not be handled by the template library at all so who's messing with the HTML?!

**Reply by JuroOravec** [reply]
Sorry I forgot to finish my thought 😅 

@oliverhaas Part of why I was asking about it is because you said that that HTML syntax could help with formatting and intellisense.

- Formatting - Totally can relate to that. Don't know if I have something misconfigured, I'm using VSCode, and I must watch out not to  have "formatting on save" enabled, because then it messes up the whole file. I tried djhtml and Djlint VSCode plugins, as mentioned in [this thread](https://www.reddit.com/r/django/comments/11v139m/django_html_formatter/).

- Intellisense - Now this is where I'm not entirely sure how it should work. Do you have in mind that the HTML attributes defined on the component would be passed through to the underlying HTML?

	E.g. that you'd have a button component that you'd use like this:
	```django
	<button class="my-btn" id="mybtn">
		Click me!
	</button>
	```

	which would be defined like so:
	```py
	class Button(Component):
		def get_context_data(self, **kwargs):
			return {
				...,
				# Pass kwargs through to allow to define HTML attributes on the component
				"attrs": kwargs,
			}
	```

	which would have template maybe like so:
	```html
	<button {% html_attrs attrs %}>
		{% slot "content" default / %}
	</button>
	```

---

On intellisense, my issue is more when it comes to validation / autocompletion of Django's features and components inputs of django_components. I'd love to have a proper [language server (LSP)](https://microsoft.github.io/language-server-protocol/), so that all Django's built-in tags / filters would be automatically autocompleted, and variables defined in loops or in `{% ... as var %}` would show up properly too.

But for me personally, working on the component library is higher priority.

---

But back to the idea of writing components as HTML. In https://github.com/EmilStenstrom/django-components/issues/644, I shared the idea of supporting plugins for django_components. If the idea is to only use different syntax for the components, but no other extra features, then this sort of template pre-processing could be implemented as a plugin. 👀

**Reply by oliverhaas** [reply]
@EmilStenstrom 
> One goal of django_components is that projects that use it should feel natural to Django-developers from the first time they see the code. I think `{% component "my_comp" abc=123 xx="{% expr '1 + 1' %}" / %}" / %}` just makes sense, but I think `<MyComp abc=123 xx="{% expr '1 + 1' %}" />` would generally confuse people, because HTML tags should not be handled by the template library at all so who's messing with the HTML?!

Looking at a typical button component in my projects (using django-components), it's basically an html button with some default styling and tiny extra features like optional icon, loading spinner, etc. So having it as `<c-button/>` still feels at least not wrong to me. 

But I'm not sure what syntax I would prefer in the long run. I just thought html-tags was a neat idea and wanted to check it out with my actual (django-)components. I expect that quite a few people will feel similar about that. It would be a pretty simple addition to django-components (an optional template loader) and I saw this discussion I wanted to share my thoughts :).

**Reply by oliverhaas** [reply]
@JuroOravec 
It's not much, but with html-tag syntax I at least get intellisense for css classes:
![image](https://github.com/user-attachments/assets/d3e73e5e-9ede-492d-85f2-55682260bdf4)

**Reply by EmilStenstrom** [reply]
@oliverhaas So the reasoning is "since my home-make components are very close to native HTML tags (your button component), seeing `<c-button>` looks natural"? The thing I think might trip people up is not understanding what's going on by looking at the code. What `<c-button>` does would then depend on which plugins are loaded, what JS libraries are loaded (it could be a normal web component), and many other things. So my reasoning is that this might feel right for someone that knows what they are doing, but is likely to confuse everyone else.

Intellisense and autoformatting for templates would be awesome :)

**Reply by JuroOravec** [reply]
Btw, @oliverhaas, you mentioned that

> We've already started implemented something along those lines in one of our projects to try it out. 

Does it mean that you are using django-cotton, or that you implemented your own script that converts the custom HTML-tags to django tags?

For context, I agree with @EmilStenstrom that this should not be in the core package. I was thinking about the plugins for django_components, and realized that they would have to operate on the level of `Template`s, not a raw string content.

So the options for you would be to:
1. Define a Django Loader that pre-processes the file content, converting the `<component />` HTML tags to django_components tags `{% component ... / %}`, and returning a `Template`

2. Define a django_components plugin that would be given a `Template`, and would do the same pre-processing on the template's source content (`Template.source`), and then re-generate the `Template.nodelist`.

Of the two, the Loader would be probably a more complex one. Because I can imagine that, to make it work generally, one might have to use similar approach to the cached Loader, nesting other loaders in it:
```py
 'loaders':[(
    'path.to.loaders.Html2DjangoLoader', [
		# Nested
       'django.template.loaders.filesystem.Loader',
       'django.template.loaders.app_directories.Loader',
		...
    ]
 )],
```

Whereas for the plugin, it would be just a matter of defining a function that modifies the `Template` in place.

So if you have the logic for converting HTML to django written down, then it should be easy to set things up once plugins are supported.

**Reply by oliverhaas** [reply]
@JuroOravec [Here](https://gist.github.com/oliverhaas/a6c2a64f02cfa6f57f23130c997e9d6a) is our proof of concept, which is basically what you said: A custom template loader which does some parsing, and then just uses django-components like normal. It's heavily inspired by django-cotton. Maybe it leads to something useful...

I think that's it for now from my end.

Thanks for the discussion, and all the good work of you two.

**Reply by EmilStenstrom** [reply]
Since this discussion happened, I've been convinced by the popularity of django-cotton that this is something that people really want. I think Juro's idea of implementing this as an optional extension is a nice way of letting people get the cookie or eat it, whichever they want.

**Reply by dalito** [reply]
Not sure if you are aware of it but [dj-angles](https://dj-angles.adamghill.com/en/stable/integrations/django-components/)  adds support for referencing components from django-components in "html-style". I haven't tested how well it works.

**Reply by EmilStenstrom** [reply]
@dalito Didn't know, cool! ping @JuroOravec
